### PR TITLE
feat: dataset metadata endpoint

### PR DESCRIPTION
**Pull request by ebezzi**
#### Reviewers
**Functional:** 
@MDunitz @leslieypark 

**Readability:** 
@seve 

---

Ticket [#59](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-explorer/59)

## Changes
- Adds a new `dataset-metadata` endpoint that calls the data portal `/collection` endpoint

## Open questions
- Does it make sense to add the new endpoint to `dataset_metadata.py` instead of `rest.py`?
- `dataset-metadata` is already used for the internal `meta=...` endpoint. Maybe we should pick a different name?

### CODE REVIEW

File: server/common/rest.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import copy
 import logging
+from server.data_common import dataset_metadata
```
Comment (leslieypark) [inline server/common/rest.py:3]: Nitpick: Move down after third party imports

**Review by leslieypark — APPROVED** [review]
🚀

File: server/tests/unit/common/test_api.py [hunk 2]
```diff
@@ -720,33 +717,86 @@ def test_dataset_does_not_exist(self):
         response = self.client.get(url)
         self.assertEqual(response.status_code, 404)
 
+
+class TestDatasetMetadata(BaseTest):
+
+    @classmethod
     @patch("server.data_common.dataset_metadata.requests.get")
```
Comment (MDunitz) [inline server/tests/unit/common/test_api.py:724]: Can this line be deleted?

File: server/tests/unit/common/test_api.py [hunk 3]
```diff
@@ -720,33 +717,86 @@ def test_dataset_does_not_exist(self):
...
-    def test_config_with_portal_metadata(self, mock_get):
-        response_body = json.dumps(
-            {
-                "collection_id": "4f098ff4-4a12-446b-a841-91ba3d8e3fa6",
-                "collection_visibility": "PUBLIC",
-                "dataset_id": "2fa37b10-ab4d-49c9-97a8-b4b3d80bf939",
-                "s3_uri": f"{FIXTURES_ROOT}/pbmc3k.cxg",
-                "tombstoned": "False",
-            }
+    def setUpClass(cls, mock_get):
```
Comment (MDunitz) [inline server/tests/unit/common/test_api.py:725]: Can you add a test where the portal response fails to ensure this fails gracefully/the explorer is still accessible if the portal is down?

Comment (ebezzi) [reply server/tests/unit/common/test_api.py:725]: Sure, that's a good idea.

File: server/data_common/dataset_metadata.py [hunk 4]
```diff
@@ -90,3 +90,36 @@ def get_dataset_metadata_for_explorer_location(dataset_explorer_location: str, a
...
+            "collection_description": res["description"],
+            "collection_contact_email": res["contact_email"],
+            "collection_contact_name": res["contact_name"],
+            "collection_links": res["links"],
+            "collection_datasets": res["datasets"],
+        }
+
+        return metadata
+    except Exception:
+        raise DatasetAccessError("Error retrieving dataset metadata")
```
Comment (MDunitz) [inline server/data_common/dataset_metadata.py:125]: I think we've used `DatasetAccessError` to indicate an issue accessing the actual dataset. Can you create a new error such as `MetadataAccessError`

**Review by MDunitz — APPROVED** [review]
One nit on the error being raised but otherwise lgtm

File: server/data_common/dataset_metadata.py [hunk 5]
```diff
@@ -90,3 +90,36 @@ def get_dataset_metadata_for_explorer_location(dataset_explorer_location: str, a
         raise DatasetNotFoundError(f"Dataset location not found for {dataset_explorer_location}")
 
     return dataset_metadata
+
+
+def get_dataset_and_collection_metadata(dataset_explorer_location: str, app_config: AppConfig):
+    data_locator_base_url = app_config.server_config.get_data_locator_api_base_url()
+    web_base_url = app_config.server_config.get_web_base_url()
+
+    base_metadata = get_dataset_metadata_for_explorer_location(dataset_explorer_location, app_config)
```
Comment (MDunitz) [inline server/data_common/dataset_metadata.py:99]: Can you use the cache here. Similar to how its done https://github.com/chanzuckerberg/single-cell-explorer/pull/64/files#diff-fbe8fa00f6e9f8f6796df09b5b4b9679567c96aa3882afe2c2bf10a3a6263d71L82

**Comment by brianraymor** [comment]
@MDunitz and @ebezzi - was there any further investigation into the [inconsistencies](https://czi-sci.slack.com/archives/C023Q1APASK/p1631302926223800?thread_ts=1631212875.184900&cid=C023Q1APASK) noted by @MillenniumFalconMechanic and myself? 

Might be good to review this while testing the new API?

**Review by MDunitz — APPROVED** [review]
lgtm

**Review by seve — APPROVED** [review]
LGMT on readability!

**Pull request merged**